File: src/main/java/com/bigdata/spark/java/sparkCore/TransformationTest.java
package com.bigdata.spark.java.sparkCore;

import com.google.common.collect.Lists;
import org.apache.spark.SparkConf;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;
import org.apache.spark.api.java.JavaSparkContext;
import scala.Tuple2;

import java.util.Iterator;
import java.util.List;

public class TransformationTest {

    public static void main(String[] args) {
        JavaSparkContext sc = getSparkContext();
        //map
//        mapOp(sc);

        //filter
//        filterOp(sc);

        //groupByKey
//        groupByKeyOp(sc);

        //reduceByKey
        reduceOp(sc);

        sc.stop();
    }

    private static void reduceOp(JavaSparkContext sc) {
        List<Tuple2<String, Integer>> list = Lists.newArrayList(
                new Tuple2<>("力量", 12),
                new Tuple2<>("敏捷", 11),
                new Tuple2<>("敏捷", 5),
                new Tuple2<>("智力", 6),
                new Tuple2<>("智力", 9)
        );

        JavaPairRDD<String, Integer> rdd = sc.parallelizePairs(list);
        rdd.reduceByKey((x, y) -> x + y).mapToPair(i -> new Tuple2<>(i._2, i._1)).sortByKey(false).foreach(e -> System.out.println(e._1 + "~" + e._2));

    }

    private static void groupByKeyOp(JavaSparkContext sc) {
        List<Tuple2<String, String>> list = Lists.newArrayList(
                new Tuple2<>("力量", "撼地神牛"),
                new Tuple2<>("敏捷", "影魔"),
                new Tuple2<>("敏捷", "幻影刺客"),
                new Tuple2<>("智力", "卡尔"),
                new Tuple2<>("智力", "冰女妹妹")
        );

        JavaPairRDD<String, String> rdd = sc.parallelizePairs(list);
        rdd.groupByKey().foreach(e -> {
            String type = e._1;
            Iterator<String> iterator = e._2.iterator();
            StringBuilder sb = new StringBuilder();
            while (iterator.hasNext()){
                sb.append(",").append(iterator.next());
            }
            if (sb.length() > 0){
                sb.deleteCharAt(0);
            }
            System.out.println("类型：" + type + " 模型：" + sb.toString());
        });

    }

    private static void filterOp(JavaSparkContext sc) {
        List<Integer> list = Lists.newArrayList(1,2,3,4,5,6,7,8,9);
        JavaRDD<Integer> rdd = sc.parallelize(list);
        rdd.filter(e -> e % 2 == 0).foreach(i -> System.out.println(i));
    }

    private static void mapOp(JavaSparkContext sc) {
        List<Integer> list = Lists.newArrayList(1,2,3,4,5,6,7,8,9);
        JavaRDD<Integer> rdd = sc.parallelize(list);
        rdd.map(e -> e * 2).foreach(i -> System.out.println(i));
    }

    public static JavaSparkContext getSparkContext() {
        SparkConf conf = new SparkConf();
        conf.setAppName("SparkArrayTest")
                .setMaster("local");
        return new JavaSparkContext(conf);
    }
}

File: spark/src/main/java/com.spark/scala/TransformationTest.scala
package com.spark.scala

import org.apache.spark.{SparkConf, SparkContext}

object TransformationTest {

  def main(args: Array[String]): Unit = {
    val sc =  getSparkContext
    //map
//    mapOp(sc)

    //filter
//    filterOp(sc)

    //groupByKey
//    groupByKey(sc)

    //reduceByKey
//    reduceByKeyOp(sc)
  }

  def mapOp(sc: SparkContext): Unit = {
    val mapRDD = sc.parallelize(Array(1,2,3,4,5,6,7,8,9))
    mapRDD.map(_ * 2).foreach(e => println(e))
  }

  def filterOp(sc: SparkContext): Unit = {
    val filterRDD = sc.parallelize(Array(1,2,3,4,5,6,7,8,9))
    filterRDD.filter(_ % 2 != 0).foreach(println(_))
  }

  /**
   * 相同的key聚合
   * @param sc
   */
  def groupByKey(sc: SparkContext): Unit = {
    val list = List(("力量", "撼地神牛"), ("敏捷", "影魔"), ("敏捷", "幻影刺客"), ("智力", "卡尔"), ("智力", "冰女妹妹"))
    val listRDD = sc.parallelize(list)
    val keyRDD = listRDD.groupByKey()
    keyRDD.foreach(e => {
      val key = e._1
      val iterable = e._2.iterator
      var pl = ""
      while (iterable.hasNext){
        pl = pl + iterable.next() + ","
      }
      var result = ""
      if(pl.length > 0){
        result = pl.dropRight(1)
      }
      println("类型:" + key + " 模型:" + result)
    })
  }

  /**
   * 将RDD中所有K,V对中K值相同的V进行合并
   * @param sc
   */
  def reduceByKeyOp(sc: SparkContext): Unit = {
    val list = List(("力量", 12), ("敏捷", 20), ("敏捷", 10), ("智力", 10), ("智力", 10))
    val listRDD = sc.parallelize(list)
    val reduceRDD = listRDD.reduceByKey(_ + _).sortBy(_._2)
    reduceRDD.foreach(e => println("类型:" + e._1 + " 人数:" + e._2))
  }

  private def getSparkContext = {
    val conf = new SparkConf()
    conf.setAppName("test")
      .setMaster("local")
    new SparkContext(conf);
  }
}
